sui_core/src/authority_server.rs

```diff
@@ -212,12 +212,24 @@ impl Validator for AuthorityServer {
     ) -> Result<tonic::Response<TransactionInfoResponse>, tonic::Status> {
         let transaction = request.into_inner();
 
-        let info = self
-            .consensus_adapter
-            .submit(&transaction)
+        // In some cases we can skip consensus for shared-object transactions: (i) we already executed
+        // the transaction, (ii) we already assigned locks to the transaction but failed to execute it.
+        // The later scenario happens when the authority missed some of the transaction's dependencies;
+        // we can thus try to re-execute it now.
+        let ConsensusTransaction::UserTransaction(certificate) = &transaction;
+        let info = match self
+            .state
+            .try_skip_consensus(certificate)
```

Comment:
You own the transaction, consider having `try_skip_consensus` take ownership of it.
See [C-CALLER-CONTROL](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control).

Reply:
thanks!

Reply:
Actually this is a good read (should have done it earlier): https://rust-lang.github.io/api-guidelines/about.html